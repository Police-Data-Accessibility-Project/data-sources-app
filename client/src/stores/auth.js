import axios from 'axios';
import { defineStore } from 'pinia';
import parseJwt from '../util/parseJwt';
import router from '../router';
import { useUserStore } from './user';

const HEADERS = {
	headers: { 'Content-Type': 'application/json' },
};
const LOGIN_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/login`;
const REFRESH_SESSION_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/refresh-session`;

export const useAuthStore = defineStore('auth', {
	state: () => ({
		userId: null,
		accessToken: {
			value: null,
			expires: Date.now(),
		},
		returnUrl: null,
	}),
	persist: true,
	actions: {
		async login(email, password) {
			const user = useUserStore();

			try {
				const response = await axios.post(
					LOGIN_URL,
					{ email, password },
					HEADERS,
				);

				// Update user store with email
				user.$patch({ email });

				this.parseTokenAndSetData(response);
				if (this.returnUrl) router.push(this.returnUrl);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		logout(isAuthRoute) {
			const user = useUserStore();

			this.$patch({
				userId: null,
				accessToken: { value: null, expires: Date.now() },
				returnUrl: null,
			});

			user.$patch({
				email: '',
			});
			if (isAuthRoute) router.push('/login');
		},

		async refreshAccessToken() {
			if (!this.$state.userId) return;
			try {
				const response = await axios.post(
					REFRESH_SESSION_URL,
					{ session_token: this.$state.accessToken.value },
					HEADERS,
				);
				return this.parseTokenAndSetData(response);
			} catch (error) {
				throw new Error(error.response?.data?.message);
			}
		},

		parseTokenAndSetData(response) {
			const token = response.data.data;
			const tokenParsed = parseJwt(token);

			this.$patch({
				userId: tokenParsed.sub,
				accessToken: {
					value: token,
					expires: new Date(tokenParsed.exp * 1000).getTime(),
				},
			});
		},
	},
});
