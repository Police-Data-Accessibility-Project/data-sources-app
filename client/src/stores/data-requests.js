import { defineStore } from 'pinia';

export const useDataRequestsStore = defineStore('data-requests', {
	state: () => ({
		/** Searches performed during session. */
		cache: {},
		/** Previous route visited - useful for determining whether we are incrementing or decrementing pages in data request by id */
		previousDataRequestRoute: null,
	}),
	getters: {
		getDataRequestFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},

	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},
		setDataRequestToCache(key, data, timestamp = new Date().getTime()) {
			this.$patch((state) => {
				state.cache[key] = {
					data,
					timestamp,
				};
			});
		},
		setPreviousDataRequestRoute(route) {
			this.$patch({
				previousDataSourceRoute: route,
			});
		},
	},
});
