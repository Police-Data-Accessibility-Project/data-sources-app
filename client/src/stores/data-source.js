import { defineStore } from 'pinia';

export const useDataSourceStore = defineStore('data-source', {
	state: () => ({
		/** Previous route visited - useful for determining whether we are incrementing or decrementing pages in data source by id */
		previousDataSourceRoute: null,
		/** Cache by ID -- useful for avoiding unnecessary API calls */
		cache: {},
	}),
	persist: {
		storage: sessionStorage,
	},
	getters: {
		getDataSourceFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},
	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},

		setPreviousDataSourceRoute(route) {
			this.$patch({
				previousDataSourceRoute: route,
			});
		},

		setDataSourceToCache(key, data) {
			this.$patch((state) => {
				state.cache[key] = {
					data,
					timestamp: new Date().getTime(),
				};
			});
		},
	},
});
