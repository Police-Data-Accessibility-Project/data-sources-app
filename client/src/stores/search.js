import { defineStore } from 'pinia';

export const useSearchStore = defineStore('search', {
	state: () => ({
		/** Searches performed during session. */
		cache: {},
		/** Needed for `NEXT` / `BACK` functionality in data source id view */
		mostRecentSearchIds: [],
		/** Needed for `NEXT` / `BACK` functionality in data requests id view */
		mostRecentRequestIds: [],
	}),
	persist: {
		storage: sessionStorage,
		pick: ['mostRecentSearchIds'],
	},
	getters: {
		getSearchFromCache: (state) => (key) => {
			if (!(key in state.cache)) {
				return null;
			}
			return state.cache[key];
		},
	},

	actions: {
		clearCache() {
			this.$patch((state) => {
				state.cache = {};
			});
		},

		setMostRecentSearchIds(ids) {
			this.$patch({
				mostRecentSearchIds: ids,
			});
		},

		setMostRecentRequestIds(ids) {
			this.$patch({
				mostRecentRequestIds: ids,
			});
		},

		setSearchToCache(key, data, timestamp = new Date().getTime()) {
			this.$patch((state) => {
				// Use object notation for setting cache data
				state.cache[key] = {
					data,
					timestamp,
				};
			});
		},
	},
});
