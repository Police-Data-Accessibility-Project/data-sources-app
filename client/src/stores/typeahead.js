import { defineStore } from 'pinia';

export const useTypeaheadStore = defineStore('typeahead', {
	state: () => ({
		/** Searches performed during session. */
		cache: {
			agencies: {},
			locations: {},
		},
	}),
	getters: {
		getTypeaheadResultsFromCache: (state) => (type, key) => {
			return state.cache?.[type]?.[key] ?? null;
		},
	},

	actions: {
		setTypeaheadResultsToCache(type, key, data) {
			this.$patch((state) => {
				// Use object notation for setting cache data
				state.cache[type][key] = {
					data,
					timestamp: new Date().getTime(),
				};
			});
		},
	},
});
