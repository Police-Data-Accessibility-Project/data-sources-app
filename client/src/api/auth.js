import axios from 'axios';
import { ENDPOINTS } from './constants';
import { useAuthStore } from '@/stores/auth';
import { useUserStore } from '@/stores/user';

const AUTH_BASE = `${import.meta.env.VITE_VUE_API_BASE_URL}/auth`;
const HEADERS = {
	'Content-Type': 'application/json',
};

export async function signUpWithEmail(email, password) {
	const user = useUserStore();

	await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.SIGNUP}`,
		{ email, password },
		{ headers: HEADERS },
	);
	// Update store with email
	user.$patch({ email });
}

export async function signInWithEmail(email, password) {
	const auth = useAuthStore();

	const response = await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.LOGIN}`,
		{ email, password },
		{
			headers: {
				...HEADERS,
				// TODO: API should require auth
				// authorization: `Basic ${import.meta.env.VITE_ADMIN_API_KEY}`,
			},
		},
	);

	auth.parseTokensAndSetData(response);
}

export async function beginOAuthLogin(redirectPath = '/sign-in') {
	const redirectTo = encodeURI(
		`${AUTH_BASE}/${ENDPOINTS.OAUTH.GITHUB}?redirect_url=${import.meta.env.VITE_VUE_APP_BASE_URL}${redirectPath}`,
	);

	window.location.href = redirectTo;
}

export async function signInWithGithub(gh_access_token) {
	const auth = useAuthStore();

	const response = await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.OAUTH.LOGIN_WITH_GITHUB}`,
		{ gh_access_token },
		{
			headers: {
				...HEADERS,
			},
		},
	);

	auth.parseTokensAndSetData(response);
	return true;
}

export async function linkAccountWithGithub(gh_access_token) {
	const { email: user_email } = useUserStore();

	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.OAUTH.LINK_TO_GITHUB}`,
		{ gh_access_token, user_email },
		{
			headers: {
				...HEADERS,
				// authorization: `Bearer ${this.$state.tokens.accessToken.value}`,
			},
		},
	);
}

export async function signOut() {
	const auth = useAuthStore();
	const user = useUserStore();

	auth.$reset();
	user.$reset();
}

export async function refreshTokens() {
	const auth = useAuthStore();
	const user = useUserStore();

	if (!user.id) return;
	try {
		const response = await axios.post(
			`${AUTH_BASE}/${ENDPOINTS.AUTH.REFRESH_SESSION}`,
			{ refresh_token: auth.$state.tokens.refreshToken.value },
			{
				headers: {
					...HEADERS,
					authorization: `Bearer ${auth.$state.tokens.accessToken.value}`,
				},
			},
		);
		return auth.parseTokensAndSetData(response);
	} catch (error) {
		console.error(error);
	}
}

export async function validateEmail(token) {
	const auth = useAuthStore();

	const response = await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.VALIDATE_EMAIL}`,
		null,
		{
			headers: {
				...HEADERS,
				Authorization: `Bearer ${token}`,
			},
		},
	);

	auth.parseTokensAndSetData(response);
}

export async function resendValidationEmail() {
	const { email } = useUserStore();

	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.RESEND_VALIDATION_EMAIL}`,
		{ email },
		{
			headers: {
				...HEADERS,
			},
		},
	);
}

export async function requestPasswordReset(email) {
	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.REQUEST_RESET_PASSWORD}`,
		{ email },
		{ headers: HEADERS },
	);
}

export async function resetPassword(password, token) {
	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.RESET_PASSWORD}`,
		{ password },
		{ headers: { ...HEADERS, Authorization: 'Bearer ' + token } },
	);
}

/**
 * @deprecated validation now done by parsing JWT directly
 */
export async function validateResetPasswordToken(token) {
	return await axios.post(
		`${AUTH_BASE}/${ENDPOINTS.AUTH.RESET_TOKEN_VALIDATION}`,
		{
			headers: {
				...HEADERS,
				Authorization: `Bearer ${token}`,
			},
		},
	);
}
