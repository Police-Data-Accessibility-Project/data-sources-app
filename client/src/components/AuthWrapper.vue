<template>
	<div id="wrapper" class="h-full w-full" v-on="handlers">
		<slot />
	</div>
</template>

<script setup>
import debounce from 'lodash/debounce';
import { useAuthStore } from '../stores/auth';
import { useRoute } from 'vue-router';
import { PRIVATE_ROUTES } from '../router';

const route = useRoute();
const auth = useAuthStore();

// Debounce func for performance
const refreshAuth = debounce(handleAuthRefresh, 350, { leading: true });

const handlers = {
	click: refreshAuth,
	keydown: refreshAuth,
	keypress: refreshAuth,
	keyup: refreshAuth,
	touchstart: refreshAuth,
	touchend: refreshAuth,
	touchcancel: refreshAuth,
	touchmove: refreshAuth,
	scroll: refreshAuth,
	submit: refreshAuth,
};

function handleAuthRefresh() {
	const isAuthRoute = PRIVATE_ROUTES.includes(route);

	const now = Date.now();
	const difference = auth.accessToken.expires - now;

	/* c8 ignore next 6 */
	if (difference > 0) {
		console.debug({
			secondsUntilRefreshOnUserActivity: (difference - 60000) / 1000,
			secondsUntilInactivityLogout: difference / 1000,
		});
	}

	// User's token is about to expire, update it.
	if (difference <= 60 * 1000 && difference > 0) {
		return auth.refreshAccessToken();
		// User's token is expired, log out.
	} else if (difference <= 0 && auth.userId) {
		return auth.logout(isAuthRoute);
	}
}
</script>
