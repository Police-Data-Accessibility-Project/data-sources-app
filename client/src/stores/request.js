import { useAuthStore } from '@/stores/auth';
import axios from 'axios';
import { defineStore } from 'pinia';

// Constants
const HEADERS_BASE = {
	authorization: `Basic ${import.meta.env.VITE_ADMIN_API_KEY}`,
	'Content-Type': 'application/json',
};
const REQUESTS_URL = `${import.meta.env.VITE_VUE_API_BASE_URL}/data-requests`;

export const useRequestStore = defineStore('request', {
	state: () => ({}),
	// persist: {
	// 	storage: sessionStorage,
	// },
	actions: {
		async createRequest(data) {
			const { tokens } = useAuthStore();
			const response = await axios.post(REQUESTS_URL, data, {
				headers: {
					...HEADERS_BASE,
					authorization: `Bearer ${tokens.accessToken.value}`,
				},
			});

			return response.data;
		},
	},
});
