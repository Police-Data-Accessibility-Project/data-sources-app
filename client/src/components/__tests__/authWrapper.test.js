import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import AuthWrapper from '../AuthWrapper.vue';
import { createTestingPinia } from '@pinia/testing';
import { useAuthStore } from '../../stores/auth';
import { nextTick } from 'vue';

let wrapper;

const NOW = Date.now();
const NOW_MINUS_THIRTY = NOW - 30 * 1000;
const NOW_PLUS_THIRTY = NOW + 30 * 1000;

describe('AuthWrapper', () => {
	beforeEach(() => {
		wrapper = mount(AuthWrapper, {
			// props: { dataSource },
			global: {
				plugins: [createTestingPinia()],
			},
		});

		vi.unstubAllGlobals();
	});

	it('renders auth wrapper', () => {
		expect(wrapper.find('[id="wrapper"]').exists()).toBe(true);
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('refreshes access token when less than 1 minute remaining before access token expiry on event', async () => {
		const auth = useAuthStore();
		auth.$patch({
			userId: 42,
			accessToken: {
				expires: NOW_PLUS_THIRTY,
			},
		});

		await wrapper.trigger('click');
		await nextTick();
		expect(auth.refreshAccessToken).toHaveBeenCalled();
	});

	it('logs user out when access token is expired on all expected events', async () => {
		const auth = useAuthStore();
		auth.$patch({
			userId: 42,
			accessToken: {
				expires: NOW_MINUS_THIRTY,
			},
		});

		await wrapper.trigger('click');
		await nextTick();
		expect(auth.logout).toHaveBeenCalled();
	});
});
