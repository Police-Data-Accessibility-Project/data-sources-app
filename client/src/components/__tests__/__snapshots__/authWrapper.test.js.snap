// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthWrapper > renders auth wrapper 1`] = `<div class="h-full w-full" id="wrapper"></div>`;
